### PR TITLE
envoy: refactor envoy agent ownership

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Move ownership of envoy agent to istio agent. This allows lifecycle control by the agent.

### CODE REVIEW

**Comment by kyessenov** [comment]
Leak checker failure seems irrelevant:
```
        internal/poll.runtime_pollWait(0x7f129b64c728, 0x72, 0x7029bd0)
        internal/poll.(*pollDesc).wait(0xc001486018, 0x72, 0x1000, 0x1000, 0xffffffffffffffff)
        internal/poll.(*pollDesc).waitRead(...)
        internal/poll.(*FD).Read(0xc001486000, 0xc000f83000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        net.(*netFD).Read(0xc001486000, 0xc000f83000, 0x1000, 0x1000, 0x0, 0x736f420, 0x0)
        net.(*conn).Read(0xc000b94128, 0xc000f83000, 0x1000, 0x1000, 0x4492fb, 0xc000cac410, 0x10)
        net/http.(*persistConn).Read(0xc000e76fc0, 0xc000f83000, 0x1000, 0x1000, 0x1, 0x0, 0xc00133fa60)
        bufio.(*Reader).fill(0xc000adb740)
        bufio.(*Reader).Peek(0xc000adb740, 0x1, 0x0, 0x1, 0x4, 0x1, 0x3)
        net/http.(*persistConn).readLoop(0xc000e76fc0)
        created by net/http.(*Transport).dialConn
        Leaked goroutine: goroutine 228 [select]:
        net/http.(*persistConn).writeLoop(0xc000e76fc0)
        created by net/http.(*Transport).dialConn
```

This PR does not create any new connections or re-orders them.

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -125,74 +122,39 @@ var (
...
-				Sidecar:           proxy.Type == model.SidecarProxy,
-			})
-
-			drainDuration, _ := types.DurationFromProto(proxyConfig.TerminationDrainDuration)
-			envoyAgent := envoy.NewAgent(envoyProxy, drainDuration)
 			// On SIGINT or SIGTERM, cancel the context, triggering a graceful shutdown
-			go cmd.WaitSignalFunc(cancel)
-
-			return envoyAgent.Run(ctx)
+			cmd.WaitSignalFunc(cancel)
```
Comment (howardjohn) [inline pilot/cmd/pilot-agent/main.go:156]: Its a bit concerning no tests fail, but I think this broke the graceful shutdown (maybe not this line exactly, but the PR).

With this PR:
```
^C2021-04-22T07:47:42.469739Z   warning envoy main      caught SIGINT
2021-04-22T07:47:42.472949Z     warning envoy config    StreamAggregatedResources gRPC config stream closed: 13,
2021-04-22T07:47:42.474382Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
2021-04-22T07:47:42.474458Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
```

Without this PR:
```
^C2021-04-22T07:48:05.557807Z   warning envoy main      caught SIGINT
2021-04-22T14:48:05.557983Z     info    Agent draining Proxy
2021-04-22T14:48:05.557983Z     info    Status server has successfully terminated
2021-04-22T14:48:05.558239Z     error   accept tcp [::]:15020: use of closed network connection
2021-04-22T07:48:05.560170Z     warning envoy config    StreamAggregatedResources gRPC config stream closed: 13,
2021-04-22T07:48:05.561390Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
2021-04-22T07:48:05.561468Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
2021-04-22T14:48:05.562976Z     info    ads     ADS: "@" .-2 terminated rpc error: code = Canceled desc = context canceled
2021-04-22T14:48:05.562982Z     info    ads     ADS: "@" .-1 terminated rpc error: code = Canceled desc = context canceled
2021-04-22T14:48:05.574072Z     info    failed draining listeners for Envoy on port 15000: Post "http://127.0.0.1:15000/drain_listeners?inboundonly&graceful": read tcp 127.0.0.1:43376->127.0.0.1:15000: read: connection reset by peer
2021-04-22T14:48:05.574139Z     warn    Error in invoking drain listeners endpoint Post "http://127.0.0.1:15000/drain_listeners?inboundonly&graceful": read tcp 127.0.0.1:43376->127.0.0.1:15000: read: connection reset by peer
2021-04-22T14:48:05.574161Z     info    Graceful termination period is 3s, starting...
2021-04-22T14:48:08.576413Z     info    Graceful termination period complete, terminating remaining proxies.
2021-04-22T14:48:08.576450Z     warn    Aborted all epochs
2021-04-22T14:48:08.576459Z     info    Agent has successfully terminated
```

This is with `terminationDrainDuration: 3s` configured

Comment (howardjohn) [reply pilot/cmd/pilot-agent/main.go:156]: oh probably since we don't block on agent.Run() completetion

Comment (kyessenov) [reply pilot/cmd/pilot-agent/main.go]: Fixed.

**Comment by kyessenov** [comment]
@howardjohn How do I debug the leaking goroutine? There is no source code reference in the trace.

**Comment by howardjohn** [comment]
I think it means some HTTP connection is not closed. Often from missing `resp.Body.Close` or `httpServer.Close()`

**Comment by kyessenov** [comment]
@howardjohn Re-ordering some code fixes the issue. So either bootstrap generation has an HTTP server / makes an HTTP connection (???) or the leak checker is flakey.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**